src/libsyntax/parse/parser.rs

```diff
@@ -4070,6 +4088,9 @@ impl<'a> Parser<'a> {
                     TraitBoundModifier::None
                 };
                 bounds.push(TraitTyParamBound(poly_trait, modifier));
+                if has_parens {
+                    self.expect(&token::CloseDelim(token::Paren))?;
+                }
             } else {
                 break
```

Comment:
This `breaks` happens only if the code is already syntactically incorrect, something like 
```rust
(Trait) + (Trait) + ($*&^#&*%%a;ch
```
so there may be no closing paren.
I can move `expect(closing_paren)` below `else { ... }` though, it should look a bit nicer.